### DISCUSSION TITLE
Passing information to and from scripts

### DISCUSSION THREAD

**Discussion by dcpurton** [neomutt/neomutt#3216] [Development]
When creating multipart attachment groups, it's often necessary to pass information to a script.

Consider a final email structure (once sent):

```
  I     1 <no description>                                                  [multipart/related, 7bit, 27K]
  I     2 ├─><no description>                                          [multipart/alternative, 7bit, 1.1K]
  I     3 │ ├─><no description>                                       [text/plain, quoted, us-ascii, 0.1K]
  I     4 │ └─><no description>                                           [text/html, quoted, utf-8, 0.7K]
  I     5 └─>neomutt-256.png                                                      [image/png, base64, 25K]
  I     6 signature.asc                                            [application/pgp-signature, 7bit, 0.8K]
```

It's quite hard to automate the creation of such an email.

Ideally we want to start with a simple structure at compose time:

```
- I     1 /tmp/neomutt-hostname-XXXX-XXXXXXX-XXXXXXXX                   [text/plain, 7bit, us-ascii, 0.1K]
  A     2 contrib/logo/neomutt-256.png                                            [image/png, base64, 19K]
```

The `text/plain` body should have some way of referencing the attached image. Let's say we use Markdown:

```
![Alt Text](cid:xxxxx)
```
Where `cid:xxxxx` refers to `neomutt-256.png`.  How to do this?

- File name?
- Attachment number?

Desired outcome is a one key macro with script to construct the final email. The macro/script needs to:

1. Parse the `text/plain` part looking for references to images.
2. Find the attachments referenced in the `text/plain` and set a `Content-ID` (default random).
3. Optionally set a `d_filename` for each image.
4. Create the multipart related group.
5. Somehow edit the the `text/plain` part to update `cid:xxxxx`.
6. Convert the `text/plain` part to `text/html` using, for example, `pandoc`.
7. Attach the `text/html` part
8. Create the `multipart/alternative` group.
9. Send the email.

*Difficult*

**Comment by flatcap** [comment]
Here's a scriptable solution, with minimal code changes...

- All attachments are given a `Content-ID`
  NeoMutt doesn't know whether they'll get used, but they're small

- `<filter-entry>` is modified to allow a `%l` lookup parameter
  If used, NeoMutt creates a temp file containing all the attachments and their `Content-ID`s.

### Filter-entry

`<filter-entry>` takes a unix command.
Adding an expando `%l` and creating a temporary file is simple enough.

**Note**:
> `<filter-entry>` overwrites the original file.
> WARNING!  You are about to overwrite file.c, continue? ([no]/yes): 

We could add some extra, "magic", expandos:

- `%g` - automatically perform a `<get-attachment>` - get a temporary copy of an attachment
- `%f` - forcibly overwrite the original file

Or we could use config instead.

### Lookup File

We want to give the user as much info as we can.
A CSV (or tab-separated) file containing:

- Content-ID
- Filename
- Description
- Mime-Type
- Content-Type
- Encoding
- Language

### Example

- Compose
- Attach: `file.txt`, `file.html`, `cat.gif`, `dog.gif`
- Group text attachments as `multipart/alternative`
- Select `file.html`
- `<filter-entry> transform.py %l %g %f`
  - NeoMutt writes lookup filee
  - NeoMutt gets a copy of the file
  - NeoMutt disables the warning message
  - NeoMutt runs `transform.py /tmp/file.html /tmp/lookup.csv`
  - Script does its magic
- Send message

**Reply by dcpurton** [reply]
> Here's a scriptable solution, with minimal code changes...
> 
> * All attachments are given a `Content-ID`
>   NeoMutt doesn't know whether they'll get used, but they're small

How short is short? Is `MUTT_RANDTAG_LEN` too long?

**Reply by flatcap** [reply]
@flatcap looks up the length: 16
Yeah, if it's good enough for the boundaries, that should be fine.

**Comment by dcpurton** [comment]
It's not difficult to modify `<filter/pipe-entry>` (and `<filter/pipe-message>`) to use Neomutt's expandos and so pass some information about the tagged email or attachment to a script. I had ago at doing this already in a separate branch, but not pushed.

One weakness of this is if multiple messages or attachments are selected and `$attach_split = no` and/or `$pipe_split = no` is set, then you will get the data for the first tagged message/attachment used for everything.

I'll make some incremental changes to `devel/multipart-related` and see how your ideas work.
